src/main/host/descriptor/socket/inet/tcp.rs

```diff
@@ -406,6 +409,287 @@ impl LegacyTcpSocket {
+                    &val,
+                    optval_ptr,
+                    optlen as usize,
+                )?;
+
+                Ok(bytes_written as libc::socklen_t)
+            }
+            (libc::SOL_TCP, libc::TCP_CONGESTION) => {
+                // the value of TCP_CA_NAME_MAX in linux
+                const CONG_NAME_MAX: usize = 16;
+
+                if optval_ptr.is_null() {
+                    return Err(Errno::EINVAL.into());
+                }
+
+                let name: *const libc::c_char =
+                    unsafe { c::tcpcong_nameStr(c::tcp_cong(self.as_legacy_tcp())) };
+                assert!(!name.is_null(), "shadow's congestion type has no name");
+                let name = unsafe { CStr::from_ptr(name) };
+                let name = name.to_bytes_with_nul();
+
+                let bytes_to_copy = *[optlen as usize, CONG_NAME_MAX, name.len()]
+                    .iter()
+                    .min()
+                    .unwrap();
+
+                let name = &name[..bytes_to_copy];
+                let optval_ptr = TypedPluginPtr::new::<u8>(optval_ptr, bytes_to_copy);
+
+                memory_manager.copy_to_ptr(optval_ptr, name)?;
+
+                // the len value returned by linux seems to be independent from the actual string length
+                Ok(std::cmp::min(optlen as usize, CONG_NAME_MAX) as libc::socklen_t)
+            }
+            (libc::SOL_SOCKET, libc::SO_SNDBUF) => {
+                let sndbuf_size: libc::c_int =
+                    unsafe { c::legacysocket_getOutputBufferSize(self.as_legacy_socket()) }
+                        .try_into()
+                        .unwrap();
+
+                let bytes_written = write_partial::<libc::c_int, _>(
+                    memory_manager,
+                    &sndbuf_size,
+                    optval_ptr,
+                    optlen as usize,
+                )?;
+
+                Ok(bytes_written as libc::socklen_t)
+            }
+            (libc::SOL_SOCKET, libc::SO_RCVBUF) => {
+                let rcvbuf_size: libc::c_int =
+                    unsafe { c::legacysocket_getInputBufferSize(self.as_legacy_socket()) }
+                        .try_into()
+                        .unwrap();
+
+                let bytes_written = write_partial::<libc::c_int, _>(
+                    memory_manager,
+                    &rcvbuf_size,
+                    optval_ptr,
+                    optlen as usize,
+                )?;
+
+                Ok(bytes_written as libc::socklen_t)
+            }
+            (libc::SOL_SOCKET, libc::SO_ERROR) => {
+                // return error for failed connect() attempts
+                let conn_err = unsafe { c::tcp_getConnectionError(self.as_legacy_tcp()) };
+
+                let error = if conn_err == -libc::ECONNRESET || conn_err == -libc::ECONNREFUSED {
+                    // result is a positive errcode
+                    -conn_err
+                } else {
```

Comment:
Why do we ignore other errors?

Reply:
No idea :) I'm just copying what the C code does:

https://github.com/shadow/shadow/blob/645a1788e09b1016e1f9bda4717d322bd4388d97/src/main/host/syscall/socket.c#L373-L386